schnauzer: add support for update repositories in China

jpculp opened the pull request:
**Description of changes:**

This adds helpers that settings generators can use to set the `updates.metadata-base-url` and `updates.targets-base-url` values to regionalized repositories. For now, this only includes China regions `cn-north-1` and `cn-northwest-1`, while current regions will continue to use the public updates repository, updates.bottlerocket.aws.

**Testing done:**

Built and launched Bottlerocket and verified the expected `updates` settings in:

- [x] us-west-2
- [x] cn-north-1
- [x] vmware (thanks, @mchaker!)

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
schnauzer: add support for update repositories in China

jpculp commented:
Added `updates.targets-base-url` migration.

jpculp commented:
Rebased.

bcressey commented on sources/models/shared-defaults/aws-tuf.toml:
```diff
@@ -0,0 +1,7 @@
+
+[metadata.settings.updates.metadata-base-url]
+setting-generator = "schnauzer settings.updates.metadata-base-url"
+template = "{{ tuf-prefix settings.aws.region }}{{ metadata-prefix settings.aws.region }}2020-07-07/{{ os.variant_id }}/{{ os.arch }}/"
```
There's what looks like an inconsistency here between `tuf-prefix` - doesn't include trailing slash? - and `metadata-prefix` - which does?

Subjectively, "without trailing slash" seems cleaner since then the final structure is more obvious in the template.

jpculp replied:
I wasn't quite sure how handlebars would react with `metadata-prefix` being an empty string, but it turns out it's fine. The weirdness here is our regional repositories will have the metadata in a `metadata` directory that isn't present in the current public repository.

jpculp commented:
Added symlinks for `k8s-1.23` variants.

jpculp commented:
- Moved the trailing `/` out of `metadata-prefix` and added it to the template instead.
- Built and launched an `aws-k8s-1.23` variant in `us-west-2` to verify there isn't a space or something in the rendered setting.

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -737,6 +749,98 @@ pub fn pause_prefix(
+        template: template_name,
+    })?;
+
+    // construct the prefix
+    let metadata_location = if TUF_BUCKET_MAP.contains_key(aws_region) {
+        "/metadata"
+    } else {
+        ""
+    };
```
nit: IMO the template would be slightly more clear if we omit the leading `/` here:
```suggestion
    // construct the prefix
    let metadata_location = if TUF_BUCKET_MAP.contains_key(aws_region) {
        "metadata"
    } else {
        ""
    };
```

So then the template would be:
```
{{ tuf-prefix settings.aws.region }}/{{ metadata-prefix settings.aws.region }}/2020-07-07/{{ os.variant_id }}/{{ os.arch }}/
```

jpculp replied:
The problem there would be regions pointing to the current public repository would render as: `https://updates.bottlerocket.aws//2020-07-07/aws-k8s-1.23/x86_64/`.

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1266,6 +1370,24 @@ fn pause_registry<S: AsRef<str>>(region: S) -> String {
+        None => (STANDARD_PARTITION),
+        Some(partition) => *partition,
+    };
+    match partition {
+        "aws-cn" => format!("https://{}.s3.{}.amazonaws.com.cn/latest", bucket, region),
+        _ => format!("https://{}.s3.{}.amazonaws.com/latest", bucket, region),
+    }
```
Per discussion, we should try to use the dual stack endpoints here for forward compatibility with IPv6-only subnets.

jpculp replied:
Rather than just specifying the bucket, I appended `.s3.dualstack` to the strings directly _(and verified it still works for cn-north-1 and cn-northwest-1)_. In the event that we add a region that does not support dual stack endpoints we can just do `<bucket_id>.s3`. We could get fancier with the map, but I wanted to keep things simple.

jpculp committed:
migrations: add targets-base-url setting-generator

AWS variants now generate the `settings.updates.targets-base-url`
setting based on region.

jpculp commented:
Tweaked the code a bit to support virtual hosted-style dual-stack s3 endpoints for forward compatibility with IPv6-only subnets.
